packages/protocol/contracts/psp/AirnodePsp.sol

```diff
@@ -0,0 +1,109 @@
+    // is good enough reason to allow PSP to reuse more generic templates with additional
+    // subscription-specific parameters.
+    struct Subscription {
+        bytes32 templateId;
+        address conditionAddress;
+        bytes4 conditionFunctionId;
+        address fulfillAddress;
+        bytes4 fulfillFunctionId;
+        bytes parameters;
+    }
+
+    IAirnodeRrp public airnodeRrp;
+
+    mapping(bytes32 => Subscription) public subscriptions;
+
+    constructor(address airnodeRrp_) {
+        airnodeRrp = IAirnodeRrp(airnodeRrp_);
+    }
+
+    // We store subscriptions in storage because we don't trust event logs to persist.
+    // A subscription should be able to live for years when left untouched.
+    function createSubscription(
+        bytes32 templateId,
+        address conditionAddress,
+        bytes4 conditionFunctionId,
+        address fulfillAddress,
+        bytes4 fulfillFunctionId,
+        bytes calldata parameters
+    ) external returns (bytes32 subscriptionId) {
+        // Subscription IDs do not need to be unique. Airnode will not serve the same
+        // sbuscription ID multiple times
+        subscriptionId = keccak256(
+            abi.encode(
+                templateId,
+                conditionAddress,
+                conditionFunctionId,
+                fulfillAddress,
+                fulfillFunctionId,
+                parameters
+            )
+        );
+        subscriptions[subscriptionId] = Subscription({
+            templateId: templateId,
+            conditionAddress: conditionAddress,
+            conditionFunctionId: conditionFunctionId,
+            fulfillAddress: fulfillAddress,
+            fulfillFunctionId: fulfillFunctionId,
+            parameters: parameters
+        });
+    }
+
+    // We are having the node fulfill through this function because we don't want the requester
+    // contract to have to import ECDSA.sol
+    // The node doesn't attempt to fulfill if `callSuccess` is false, logs the revert string internally.
+    function fulfill(
+        bytes32 subscriptionId,
+        uint256 timestamp,
+        bytes calldata data,
+        bytes calldata signature
+    ) external returns (bool callSuccess, bytes memory callData) {
+        Subscription storage subscription = subscriptions[subscriptionId];
+        // Note that we are not verifying that msg.sender is a pre-determined `sponsorWallet`.
+        // This because one cannot "botch" (e.g., by specifying a low gas limit) a PSP fulfillment
+        // like an RRP fulfillment (because previous failed fulfillments don't prevent future
+        // successful ones).
+        (address airnode, , ) = airnodeRrp.templates(subscription.templateId);
+        // It is a concern for someone to have the API provider sign a response, to be used much
+        // later (for example, would be very problematic for asset price data feeds). Therefore,
+        // we add the timestamp to the hash to be signed. 
```

Comment:
The idea is that a requester will compare the `timestamp` with the `now` in solidity? and decide depending on that?

I wonder whether this achieves anything the API provider can sign it with a future timestamp and there is no way for requester to check if the timestamp is indeed valid... Maybe you could ask the API provider to specify the hash and the block number of the latest mined block - since that can be at least verified by the requester.

Reply:
> API provider can sign it with a future timestamp
Good point, though I was mostly thinking about transactions stuck in the mempool more than a malicious API provider, they can straight out misreport anyway

> specify the hash and the block number of the latest mined block
You can only get the hash of one of the 256 latest blocks in contract, so this would be pretty limiting (and probably not future-proof). Also we would have to make more blockchain provider calls by the node. So this is creating more problems than it solves imo (mostly because it doesn't protect against all kinds of attacks that an API provider performs).